contracts/contracts/coordination/BetaProgramInitiator.sol

```diff
@@ -0,0 +1,162 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+
+pragma solidity ^0.8.0;
+
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+
+import "./Coordinator.sol";
+
+contract BetaProgramInitiator {
+    using SafeERC20 for IERC20;
+
+    event RequestRegistered(
+        address indexed sender,
+        uint256 indexed requestIndex,
+        address[] providers,
+        address authority,
+        uint32 duration,
+        IEncryptionAuthorizer accessController,
+        uint256 payment
+    );
+
+    event RequestCanceled(address indexed sender, uint256 indexed requestIndex);
+
+    event RequestExecuted(uint256 indexed requestIndex, uint256 indexed ritualId);
+
+    event FailedRequestRefunded(uint256 indexed requestIndex, uint256 refundAmount);
+
+    struct InitiationRequest {
+        address[] providers;
+        address authority;
+        uint32 duration;
+        IEncryptionAuthorizer accessController;
+        address sender;
+        uint32 ritualId;
+        uint256 payment;
+    }
+
+    uint32 public constant NO_RITUAL = type(uint32).max;
+
+    Coordinator public immutable coordinator;
+    IERC20 public immutable currency;
+    address public immutable executor; // TODO transferable role?
+
+    InitiationRequest[] public requests;
+
+    constructor(Coordinator _coordinator, address _executor) {
+        require(_executor != address(0), "Invalid parameters");
+        coordinator = _coordinator;
+        currency = coordinator.currency();
+        executor = _executor;
+    }
+
+    function getRequestsLength() external view returns (uint256) {
+        return requests.length;
+    }
+
+    function getProviders(uint256 requestIndex) external view returns (address[] memory) {
+        InitiationRequest storage request = requests[requestIndex];
+        return request.providers;
+    }
+
+    function registerInitiationRequest(
+        address[] calldata providers,
+        address authority,
+        uint32 duration,
+        IEncryptionAuthorizer accessController
+    ) external returns (uint256 requestIndex) {
+        uint256 ritualCost = coordinator.getRitualInitiationCost(providers, duration);
+
+        requestIndex = requests.length;
+        InitiationRequest storage request = requests.push();
+        request.providers = providers;
+        request.authority = authority;
+        request.duration = duration;
+        request.accessController = accessController;
+        request.sender = msg.sender;
+        request.ritualId = NO_RITUAL;
+        request.payment = ritualCost;
+
+        emit RequestRegistered(
+            msg.sender,
+            requestIndex,
+            providers,
+            authority,
+            duration,
+            accessController,
+            ritualCost
+        );
+        currency.safeTransferFrom(msg.sender, address(this), ritualCost);
```

Comment:
What do you think if funds are transferred later, when the request is executed?

Reply:
Since anyone can register a request - transferring the funds upfront could be a good anti-griefing mechanism.

Reply:
Good point